src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/AssemblyIdentityMustMatch.cs

```diff
@@ -31,8 +35,9 @@ private void RunOnAssemblySymbol(IAssemblySymbol left, IAssemblySymbol right, st
                 return;
             }
 
-            AssemblyIdentity leftIdentity = left.Identity;
-            AssemblyIdentity rightIdentity = right.Identity;
+            // At this point, left and right are both not null.
```

Comment:
NIT: Not technically true, you could have an edge case where both are null

Reply:
Interesting. That would result in a NullReferenceException today. Do you have an example of such an edge case?

Reply:
AFAIK this shouldn't happen unless there is a bug with one of the mappers, but IMO it is better to have a Debug.Assert with a searchable and diagnosable error message than just letting it null-ref by adding a `!` in order to get rid of the nullability warning.

Reply:
> than just letting it null-ref by adding a !

FWIW, this change didn't modify the existing behavior. The `!` bang operator just silences the now emitted nullability error. I agree that a Debug Assert would make sense, but IMHO it would make more sense to not have that in every rule but at the place where the mapped values are passed into the rule (RuleRunner). For the rule itself ideally we would be able to express that both parameters can't be null via i.e. the following (which AFAIK doesn't exist in the BCL today):

```c# 
Rule([NotNullWhenParameterIsNull(nameof(right)) IAssemblySymbol? left, [NotNullWhenParameterIsNull(nameof(left)) IAssemblySymbol right, ...)
```